pkg/controller/petset/pet_set.go

```diff
@@ -0,0 +1,342 @@
+package petset
+
+import (
+	"fmt"
+	"reflect"
+	"sort"
+	"time"
+
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/api/unversioned"
+	"k8s.io/kubernetes/pkg/apis/apps"
+	"k8s.io/kubernetes/pkg/client/cache"
+	"k8s.io/kubernetes/pkg/client/record"
+	client "k8s.io/kubernetes/pkg/client/unversioned"
+	"k8s.io/kubernetes/pkg/controller"
+	"k8s.io/kubernetes/pkg/controller/framework"
+	"k8s.io/kubernetes/pkg/runtime"
+	utilruntime "k8s.io/kubernetes/pkg/util/runtime"
+	"k8s.io/kubernetes/pkg/util/wait"
+	"k8s.io/kubernetes/pkg/util/workqueue"
+	"k8s.io/kubernetes/pkg/watch"
+
+	"github.com/golang/glog"
+)
+
+const (
+	// Time to sleep before polling to see if the pod cache has synced.
+	PodStoreSyncedPollPeriod = 100 * time.Millisecond
+	// number of retries for a status update.
+	statusUpdateRetries = 2
+	// period to relist petsets
+	// TODO: Increase this, currently we don't actually created pods, so we need to force relist.
+	petSetResyncPeriod = 10 * time.Second
+)
+
+type PetSetController struct {
+	kubeClient *client.Client
+
+	// newSyncer returns an interface capable of syncing a single pet.
+	// Abstracted out for testing.
+	newSyncer func(*pcb) *petSyncer
+	podStore  cache.StoreToPodLister
+
+	// podStoreSynced returns true if the pod store has synced at least once.
+	podStoreSynced func() bool
+	// Watches changes to all pods.
+	podController framework.ControllerInterface
+
+	// A store of PetSets, populated by the psController.
+	psStore cache.StoreToPetSetLister
+	// Watches changes to all PetSets.
+	psController *framework.Controller
+
+	// A store of the 1 unhealthy pet blocking progress for a given ps
+	blockingPetStore *unHealthyPetTracker
+
+	// Controllers that need to be synced.
+	queue *workqueue.Type
+
+	// To allow injection of syncPetSet for testing.
+	syncHandler func(psKey string) []error
+}
+
+// NewPetSetController creates a new petset controller.
+func NewPetSetController(kubeClient *client.Client, resyncPeriod time.Duration) *PetSetController {
+
+	eventBroadcaster := record.NewBroadcaster()
+	eventBroadcaster.StartLogging(glog.Infof)
+	eventBroadcaster.StartRecordingToSink(kubeClient.Events(""))
+	recorder := eventBroadcaster.NewRecorder(api.EventSource{Component: "petset"})
+	pc := &apiServerPetClient{kubeClient, recorder, &defaultPetHealthChecker{}}
+
+	psc := &PetSetController{
+		kubeClient:       kubeClient,
+		blockingPetStore: newUnHealthyPetTracker(),
+		newSyncer: func(blockingPet *pcb) *petSyncer {
+			return &petSyncer{pc, blockingPet}
+		},
+		queue: workqueue.New(),
+	}
+
+	// TODO: Swap this with the shared informer.
+	psc.podStore.Store, psc.podController = framework.NewInformer(
```

Comment:
the reasons I didn't was somewhat artificial. the clientset (that the shread informer constructor takes) doesn't understand how to read local kubeconfig, which is really useful for testing the controller through a main(). But I've swapped it, for greater justice. I can debug clientset later.